src/linear_algebra/basic.lean

```diff
@@ -127,6 +127,19 @@ ext $ assume b, rfl
   p.subtype.comp (cod_restrict p f h) = f :=
 ext $ assume b, rfl
 
+/-- Restrict domain and codomain of an endomorphism. -/
+def restrict (f : M →ₗ[R] M) (p : submodule R M) (hf : ∀ x ∈ p, f x ∈ p) : p →ₗ[R] p :=
```

Comment:
Two questions:
 * Is there a reason that `f` is specifically an endomorphism? The definition should still make sense if `q : submodule R M₂` and `hf : ∀ x ∈ p, f x ∈ q`.
 * Is there a reason that `p` is an explicit variable? It should be inferrable from `hf`.

So my first choice would be something like:
```lean
def restrict (f : M →ₗ[R] N) {p : submodule R M} {q : submodule R N} (hf : ∀ x ∈ p, f x ∈ p) : p →ₗ[R] p :=
```

Reply:
My thinking was that this more general version of `restrict` that you propose is probably not going to be necessary very often and that when it is necessary, `(f.cod_restrict p hf).dom_restrict q` might be the cleaner option. For endomorphisms, restricting both at once makes more sense because you put an endomorphism in and you get an endomorphism out. But I am open to generalizing it if you'd prefer.

Reply:
I'm not so sure whether the general version is that unusual, but I'm OK with leaving it to future users to implement it when they need it.

What do you think about having `p` as an implicit variable?

Reply:
Oh, sorry, I forgot about that. I changed it to implicit now. Unfortunately, that makes it inconsistent with the signature of `cod_restrict`, but I guess it's actually `cod_restrict` that should be changed, too?